### DISCUSSION TITLE
Surface book 3, graphic card not working, Ubuntu 24.04

### DISCUSSION THREAD

**Discussion by Ak00072** [linux-surface/linux-surface#1514] [General]
Hello, I have a surface book 3 equipped with a GTX1660ti graphic card, I recently downloaded Linux hoping to make it my main operating system. 
Ubuntu 24.04 LTS.
When i booted my device with linux, the nvidia graphic card was detectable and running and functional, but when I run the surface linux kernel, the system is unable to detect the graphic card, it's not visible on Nvidia X server  setting, or under system details, I also tried running the command nvidia-smi and it outputs "NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running." I have tried many different methods using chatgpt, nothing worked, only way was to uninstall the surface 6.9.3 kernel, but that would mean that i lose touch functionality and other drivers provided by the surface kernel. 
hopefully someone has a solution out `there` 
when I run : nvidia-settings
 OUTPUT= (nvidia-settings:4154): GLib-GObject-CRITICAL **: 20:10:20.104: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
** (nvidia-settings:4154): CRITICAL **: 20:10:20.108: ctk_powermode_new: assertion '(ctrl_target != NULL) && (ctrl_target->h != NULL)' failed
ERROR: nvidia-settings could not find the registry key file or the X server is
       not accessible. This file should have been installed along with this
       driver at
       /usr/share/nvidia/nvidia-application-profiles-key-documentation. The
       application profiles will continue to work, but values cannot be
       prepopulated or validated, and will not be listed in the help text.
       Please see the README for possible values and descriptions.
sudo lshw -C display:
OUTPUT= *-display UNCLAIMED
       description: 3D controller
       product: TU116M [GeForce GTX 1660 Ti Mobile]
       vendor: NVIDIA Corporation
       physical id: 0
       bus info: pci@0000:02:00.0
       version: a1
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress cap_list
       configuration: latency=0
       resources: iomemory:600-5ff iomemory:600-5ff memory:a0000000-a0ffffff memory:6000000000-600fffffff memory:6010000000-6011ffffff ioport:3000(size=128) memory:a1000000-a107ffff

**Comment by JustasStankunas** [comment]
I am sitting with the same headache right now (surface book 2). Just tried adding System76 repos and installing their nvidia driver version, but it looks that it has same problems. 
@Ak00072 did you managed to find a solution? Have you tried downgrading surface kernel version? (I will try this next)

**Reply by codekoenig** [reply]
Same problem here. Was running for a while with the default kernel and Nvidia dGPU was working great, but I had major issues with CPU throttling with the default kernel so I switched to the linux-surface-kernel ... the massive CPU throttling has gone, performance is great ... but the Nvidia card disappeared. 

Also I can hear the dGPU fan running so the GPU is on and drawing power, which is especially bad as this costs a lot of battery. 

lspci | grep -i nvidia shows me the system sees the dGPU ... but that's it.

**Edit:** Found a working fix to at least prevent the dGPU from running and drawing power all the time - executing the following command (found on https://github.com/linux-surface/surface-hotplug/wiki/Runtime-Power-Management):

`echo auto | sudo tee /sys/bus/pci/devices/<dgpu>/power/control`

seems to bring the dGPU into suspended mode and the fan turns off. Now it's at least not drawing any unneccesary power. See the linked article for more details.

**Reply by codekoenig** [reply]
Ok guys with some more digging i figured out the needed steps to make it work:

***⚠️ Before you begin:***
Make sure **Secure Boot is disabled** on your Surface (unsigned DKMS modules will otherwise be blocked).

> If you need Secure Boot on, you can follow Ubuntu’s MOK workflow to self-sign the NVIDIA modules after building them.

It turns out the root issue is that when you switch to the Surface-patched kernel, your NVIDIA driver’s DKMS modules aren’t being automatically built and loaded for that new kernel. What we’re doing here is manually installing the matching DKMS source, building the NVIDIA kernel module against the Surface kernel headers, and then activating it via Ubuntu’s PRIME system.

I added everything's necessary to achieve the desired result, but ofc some things might be already present on your system (like linux-headers-surface).

---

### 1. Re-add the Surface repo & install matching headers

```bash
curl -sSL https://raw.githubusercontent.com/linux-surface/linux-surface/master/install.sh | sudo bash
sudo apt update
sudo apt install linux-headers-surface
```

This ensures your Surface kernel headers are present so DKMS can build against them.

### 2. Install the NVIDIA DKMS source package

Pick the package matching your driver series (e.g. 570):

```bash
sudo apt install nvidia-dkms-570
```

That drops the `/usr/src/nvidia-570.xxx/` tree containing `dkms.conf`.

### 3. Build & install the NVIDIA module via DKMS

Force DKMS to compile and install the module for your current Surface kernel:

```bash
sudo dkms autoinstall
```

You should see it compile and report “installing to kernel <your-surface-kernel>”.

### 4. Update initramfs & reboot

```bash
sudo update-initramfs -u
sudo reboot
```

This packs the newly built `nvidia.ko` into your boot image.

### 5. Switch PRIME mode

* **On-demand** (default Intel, prefix `prime-run` for dGPU):

  ```bash
  sudo prime-select on-demand
  sudo reboot
  ```
* **Always NVIDIA** (full dGPU mode):

  ```bash
  sudo prime-select nvidia
  sudo reboot
  ```

### 6. Verify your dGPU’s back

After logging in:

```bash
prime-select query    # should show “on-demand” or “nvidia”
nvidia-smi            # lists GPU and driver version
```

---

#### Quick FAQs

* **DKMS “installed” but no GPU?**
  Double-check your `prime-select` setting.

* **Need Secure Boot on?**
  After step 3, generate and enroll a MOK key, then use the kernel’s `scripts/sign-file` to sign `nvidia.ko`.

* **To go back to integrated mode and save battery:**

  ```bash
  sudo prime-select intel
  sudo reboot
  ```
